server/player/player.go

```diff
@@ -367,6 +368,37 @@ func (p *Player) Transfer(address string) (err error) {
 	return
 }
 
+// SetHeldSlot updates the held slot of the player with the slot passed. It also verifies that the item within the spot
+// is the correct item.
+// This invokes the HandleHeldSlotChange handler.
+func (p *Player) SetHeldSlot(slot int, expected item.Stack) error {
+	if slot > 8 {
+		return fmt.Errorf("new held slot exceeds hotbar range 0-8: slot is %v", slot)
+	}
+
+	if p.heldSlot.Swap(uint32(slot)) == uint32(slot) {
+		return nil
+	}
+
+	p.ReleaseItem()
+
+	clientSideItem := expected
+	actual, _ := p.inv.Item(slot)
+
+	if !clientSideItem.Equal(actual) {
+		return fmt.Errorf("client side item isn't equal to expected slot item")
```

Comment:
This code no longer belongs here now that it's a player method, should be moved to packet handling (check a0bb8dde26bc089aac89d2625cb45a8a1dfe6523).